### ISSUE TITLE
fixfiles : why does -C PREVIOUS_FILECONTEXT tries to exclude non-root FS ?

### ISSUE THREAD

**Issue by cedricbu** [SELinuxProject/selinux#263]
Hello,

When running `fixfiles -C <PREVIOUS_FILECONTEXT>`, some directories, namely `/proc` `/dev` `/run` `/mnt` `/var/tmp` `/var/lib/BackupPC` `/home` `/tmp` `/dev` are excluded from relabelling.

I suspect it might be a relic of an old restorcon behavior, that prevented it to run on incompatible File Systems (that don't accept extended attributes) [see https://github.com/SELinuxProject/selinux/commit/2d0c192355ae836ffe047cacfc7c25e146f1fccb]
 
`fixfiles -C` is used used by package managers (at least in RHEL and Fedora) during an update, probably to optimize the relabelling time. However, this may lead to inconsistencies, encouraging for a full `fixfiles onboot` when new policies make changes in these excluded directories.

If the above suspicions are correct, are the static exclusion list still required ? can they possibly be removed ?

Note: 2 bits of codes seem to be aiming at excluding a static directory list :

```
159 #
160 # Compare PREVious File Context to currently installed File Context and
161 # run restorecon on all files affected by the differences.
162 #
163 diff_filecontext() {
164 EXCLUDEDIRS="`exclude_dirs_from_relabelling`"
165 for i in /sys /proc /dev /run /mnt /var/tmp /var/lib/BackupPC /home /tmp /dev; do
166     [ -e $i ]  && EXCLUDEDIRS="${EXCLUDEDIRS} -e $i";
167 done
168 LogExcluded
169 
170 if [ -f ${PREFC} -a -x /usr/bin/diff ]; then
171     TEMPFILE=`mktemp ${FC}.XXXXXXXXXX`
172     test -z "$TEMPFILE" && exit
173     PREFCTEMPFILE=`mktemp ${PREFC}.XXXXXXXXXX`
174     sed -r -e 's,:s0, ,g' $PREFC | sort -u > ${PREFCTEMPFILE}
175     sed -r -e 's,:s0, ,g' $FC | sort -u | \
176     /usr/bin/diff -b ${PREFCTEMPFILE} - | \
177         grep '^[<>]'|cut -c3-| grep ^/ | \
178         egrep -v '(^/home|^/root|^/tmp|^/dev)' |\
179     sed -r -e 's,[[:blank:]].*,,g' \
```

Lines 165-167 and 178 seems to exclude these directories.

**Comment by stephensmalley** [comment]
@rhatdan PTAL.  There is a deep magic here that I think came from Dan Walsh who I just cc'd.  I'd be glad to get rid of these hardcoded lists as long as it doesn't break anything.

**Comment by stephensmalley** [comment]
Also note that we prefer such questions to be taken to the selinux mailing list, selinux at vger.kernel.org, http://vger.kernel.org/vger-lists.html#selinux

**Comment by rhatdan** [comment]
These directories, are defined as <<none>> or are system directories, which means that restorecon is just going to waste time in them and then do nothing.  
The bottom line was to speed up the time it takes to relabel the entire system, or if relabeling was done in a post install to also eliminate useless directories.

Theoretically the script could be made smarter to look for directories without labels and for file systems that do not support labeling, to exclude.

**Comment by dlippold** [comment]
But in a recently applied correction of the SELinux rules ( https://github.com/fedora-selinux/selinux-policy/commit/71e1989028802c7875d3436fd3966c587fa383fb ) files in the home directories (in `/home/*/.config/Yubico/`) were affected. Because of the addition of `/home` to `EXCLUDEDIRS` the SELinux context for these files was not corrected and therefore a bug caused by the missing SELinux rules ( https://bugzilla.redhat.com/show_bug.cgi?id=1860888 ) was and is not corrected. From my point of view at least `/home` should not be added to `EXCLUDEDIRS`.

**Comment by cedricbu** [comment]
Just as a general idea :
Would replacing the static list of path in the `for` loop by getting the output of a command such as `findmnt --list --noheadings --output TARGET --options noseclabel` be a good start for a replacement ?
(of course, code will need to take spaces in the path into account, and I am not sure what is supposed to happen submounts, for example, /x is mounted without SELinux support, but /x/y/z is)

**Comment by rhatdan** [comment]
Yes, I would start with that.  And I would not worry about submounts.

**Comment by cedricbu** [comment]
Actually, if we trust the description of restorecon's man page, that command appear to already do a similar filtering by default (unless `-m` is in use, which I believe is not the case). So `fixfiles` adding the very same paths to the exclusion list feels redundant.

```
NAME
       restorecon - restore file(s) default SELinux security contexts.
[...]
       -m     do not read /proc/mounts to obtain a list of non-seclabel mounts to be excluded from relabeling checks.  Setting this option is useful where  there  is  a
              non-seclabel fs mounted with a seclabel fs mounted on a directory below this.
```

Is there some sort of standard test that would measure the time difference if we remove the filtering done by `fixfiles`  (i.e.: lines 165-167 and 178 from the original comment) ?

**Comment by stephensmalley** [comment]
As a reminder, these discussions should occur on the selinux@vger.kernel.org mailing list, along with any actual patch submissions.  And yes, restorecon automatically excludes non-seclabel mounts by default these days (but that wasn't always the case since seclabel wasn't always present).

**Comment by cedricbu** [comment]
ack, np.
I've sent an e-mail to the list.
